src/interchange/src/protobuf.rs

```diff
@@ -30,92 +30,106 @@ fn proto_message_name(message_name: &str) -> String {
-                    validate_proto_field_resolved(seen_messages, &f, descriptors)?;
-                }
-                seen_messages.remove(m.name());
+) -> Result<ScalarType> {
+    Ok(match field.field_type(descriptors) {
+        FieldType::Bool => ScalarType::Bool,
+        FieldType::Int32 | FieldType::SInt32 | FieldType::SFixed32 => ScalarType::Int32,
+        FieldType::Int64 | FieldType::SInt64 | FieldType::SFixed64 => ScalarType::Int64,
+        FieldType::Enum(_) => ScalarType::String,
+        FieldType::Float => ScalarType::Float32,
+        FieldType::Double => ScalarType::Float64,
+        FieldType::UInt32 => bail!("Protobuf type \"uint32\" is not supported"),
+        FieldType::UInt64 => bail!("Protobuf type \"uint64\" is not supported"),
+        FieldType::Fixed32 => bail!("Protobuf type \"fixed32\" is not supported"),
+        FieldType::Fixed64 => bail!("Protobuf type \"fixed64\" is not supported"),
+        FieldType::String => ScalarType::String,
+        FieldType::Bytes => ScalarType::Bytes,
+        FieldType::Message(m) => {
+            if seen_messages.contains(m.name()) {
+                bail!("Recursive types are not supported: {}", m.name());
             }
-            FieldType::Bytes => {
-                bail!("Arrays or nested messages with bytes objects are not currently supported")
+            seen_messages.insert(m.name());
+            let mut fields = Vec::with_capacity(m.fields().len());
+            for field in m.fields() {
+                let column_name = ColumnName::from(field.name());
+                let scalar_type =
+                    derive_scalar_type_from_proto_field(seen_messages, field, descriptors, true)?;
+                let nullable = match field.field_label() {
+                    FieldLabel::Optional => true,
+                    FieldLabel::Repeated | FieldLabel::Required => false,
+                };
+                let column_type = ColumnType {
+                    scalar_type,
+                    nullable,
+                };
+                fields.push((column_name, column_type))
             }
-            FieldType::Group => bail!("Unions are currently not supported"),
-            FieldType::UnresolvedMessage(a) => bail!("Nested message type {} unresolved", a),
-            FieldType::UnresolvedEnum(e) => bail!("Unresolved enum type {}", e),
-        },
-    }
+            seen_messages.remove(m.name());
+            ScalarType::Record {
+                fields,
+                custom_oid: None,
+                custom_name: None,
+            }
+        }
+        FieldType::Group => bail!("Unions are currently not supported"),
+        FieldType::UnresolvedMessage(m) => bail!("Unresolved message {} not supported", m),
+        FieldType::UnresolvedEnum(e) => bail!("Unresolved enum {} not supported", e),
+    })
+}
 
-    Ok(())
+// Cast all nested numeric types to a float.
+fn derive_nested_scalar_type<'a>(field_type: &FieldType) -> Option<ScalarType> {
```

Comment:
Can you add a comment explaining why casting ints to floats makes sense? I think it made sense for JSON because json is definted to only have floats, but with records I'd expect to be able to do better?

Reply:
We can! I just did it wrong the first time I tried, it turns out. Fixed!